### DISCUSSION TITLE
Task fails on update

### DISCUSSION THREAD

**Discussion by tpluscode** [copier-org/copier#586] [Q&A] [answered]
I think I'm confused about how tasks work

I added the following task which adds a dependency to the local `package.json`

```yaml
_tasks:
  - "{% if changesets %}npx add-dependencies --dev @changesets/cli{% endif %}"
```

This is a partial template and it works fine on the first time I run `copier copy`. The dependency is added as expected.

However, when I try to `copier update`, I find one problem and one surprise.

The surprise is that the task is immediately executed, before copier even asks the questions.

More importantly though, the command fails. The error is `Could not read from ./package.json` and while I haven't yet figured out exactly why, the same command works when executed directly. This leads me to think that copier maybe runs this in a different working directory?

**Comment by pawamoy** [comment, marked as answer]
When updating, Copier first regenerates a fresh project using the current template version, in a temporary folder. It executes the tasks in there, this is what you see in the output, even before the questions. Then it computes the diff between the fresh version and your current version. Then it updates your current project in-place with the new template version. Finally it reapplies the diff it computed earlier. See https://copier.readthedocs.io/en/latest/updating/#how-the-update-works

**Reply by tpluscode** [reply]
Ah of course. This makes total sense. The `package.json` is not present in the temp dir so the script has no way but fail.

I will have to guard against that possibility.

Thank you
